File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/auth/application/query/CredentialQuery.java
package com.example.accountservice.auth.application.query;

import com.example.accountservice.auth.domain.CredentialFindQuery;
import com.example.accountservice.auth.domain.CredentialsWithCount;

/**
 * The read-only lookup contract for Credential (the same role as AccountQuery/CardQuery in
 * cqrs-pattern.md). Both Command Services — SignUpService (checking for a duplicate ID) and
 * SignInService (looking up the stored hash for password verification) — use only this Query. Since
 * a Credential is never modified after creation, a read-only Query suffices for both use cases
 * instead of the write-side CredentialRepository.
 */
public interface CredentialQuery {
    CredentialsWithCount findCredentials(CredentialFindQuery query);
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/auth/application/query/CredentialQuery.kt
package com.example.accountservice.auth.application.query

import com.example.accountservice.auth.domain.Credential
import com.example.accountservice.auth.domain.CredentialFindQuery

/**
 * The Credential read-only port — both checking for a duplicate user ID (SignUpService) and looking up
 * the hash for password verification (SignInService) go through this interface. SignInService never
 * saves credentials (it only reads then verifies), so it doesn't access the write model
 * [com.example.accountservice.auth.domain.CredentialRepository].
 *
 * The query method follows the root `repository-pattern.md`'s `find<Noun>s` naming convention — rather
 * than having a dedicated single-record lookup method (like `findByUserId`), it's handled with
 * `CredentialFindQuery(take = 1, userId = ...)` + `.first.firstOrNull()` (the same pattern as
 * AccountQuery/CardQuery/PaymentQuery).
 */
interface CredentialQuery {
    fun findCredentials(query: CredentialFindQuery): Pair<List<Credential>, Long>
}
